crypto/src/hash.rs

```diff
@@ -165,6 +158,33 @@ impl Encode for Hash {
     }
 }
 
+impl FromStr for Hash {
+    type Err = Error;
+
+    fn from_str(key: &str) -> Result<Self, Self::Err> {
+        let hash: [u8; Self::LENGTH] = hex_decode(key)?.try_into().map_err(|hash_vec| {
+            Error::Parse(format!(
+                "Unable to parse {hash_vec:?} as [u8; {}]",
+                Self::LENGTH
+            ))
+        })?;
+
+        Hash::is_lsb_1(&hash)
+            .then_some(hash)
+            .ok_or_else(|| Error::Parse("expect least significant bit of hash to be 1".to_owned()))
+            .map(Self::prehashed)
+    }
+}
+
+impl<'de> Deserialize<'de> for Hash {
```

Comment:
Can be derived with [`serde_with`](https://docs.rs/serde_with/latest/serde_with/derive.DeserializeFromStr.html)